src/hera/_cli/generate/hera.py

```diff
@@ -0,0 +1,434 @@
+import os
+import sys
+from collections import ChainMap
+from dataclasses import (
+    dataclass,
+    field as dc_field,
+)
+from pathlib import Path
+from typing import Any, Generator, Iterable, List, Optional, Tuple, Type, Union, cast, get_args
+
+import yaml
+
+from hera._cli.base import GeneratePython
+from hera._cli.generate.util import YAML_EXTENSIONS, expand_paths, filter_paths
+from hera.shared._pydantic import BaseModel
+from hera.shared._type_util import get_annotated_metadata, is_optionally_subtype, unwrap_annotation
+from hera.workflows._meta_mixins import ModelMapperMixin, _get_model_attr
+from hera.workflows.cluster_workflow_template import ClusterWorkflowTemplate
+from hera.workflows.container import Container
+from hera.workflows.container_set import ContainerNode, ContainerSet
+from hera.workflows.cron_workflow import CronWorkflow
+from hera.workflows.dag import DAG
+from hera.workflows.data import Data
+from hera.workflows.http_template import HTTP
+from hera.workflows.models import (
+    ClusterWorkflowTemplate as _ModelClusterWorkflowTemplate,
+    CronWorkflow as _ModelCronWorkflow,
+    Metadata,
+    Template,
+    Workflow as _ModelWorkflow,
+    WorkflowTemplate as _ModelWorkflowTemplate,
+)
+from hera.workflows.resource import Resource
+from hera.workflows.script import Script
+from hera.workflows.steps import Step, Steps
+from hera.workflows.suspend import Suspend
+from hera.workflows.task import Task
+from hera.workflows.workflow import Workflow
+from hera.workflows.workflow_template import WorkflowTemplate
+
+try:
+    from inspect import get_annotations  # type: ignore
+except ImportError:
+    from hera.shared._inspect import get_annotations  # type: ignore
+
+DEFAULT_EXTENSION = ".py"
+
+ModelWorkflow = Union[
+    _ModelWorkflow,
+    _ModelWorkflowTemplate,
+    _ModelClusterWorkflowTemplate,
+    _ModelCronWorkflow,
+]
+
+
+@dataclass
+class FileBuilder:
+    hera_imports: list[str] = dc_field(default_factory=list)
+    model_imports: list[str] = dc_field(default_factory=list)
+    class_def: list[str] = dc_field(default_factory=list)
+    context_def: list[str] = dc_field(default_factory=list)
+
+
+def generate_workflow(options: GeneratePython):
+    """Generate Python (Hera) Workflow definitions from YAML definitions.
+
+    If the provided path is a folder, generates python for all yaml files containing `Workflow`s
+    in that folder
+    """
+    paths = sorted(expand_paths(options.from_, YAML_EXTENSIONS, recursive=options.recursive))
+
+    # Generate a collection of source file paths and their resultant python.
+    path_to_output: list[tuple[str, str]] = []
+    for path in filter_paths(paths, includes=options.include, excludes=options.exclude):
+        python_outputs = []
+        for workflow in load_yaml_workflows(path):
+            python_outputs.append(workflow_to_python(workflow))
```

Comment:
We probably want to add the filename and workflow name to any validation error we raise here, otherwise users are going to have a hard time knowing where errors came from.

Reply:
This function was mostly a copy of `generate_yaml` from src/hera/_cli/generate/yaml.py, which I was intending to refactor with this change but wanted to get it working first. I would prefer to do that in a follow up minimal-functional-changes PR (but I agree telling the user what the problem was is a good idea, which is not done for YAML conversion).